File: parse.js
const fs = require('fs');

class TMLParser {
    constructor(tmlInput = null, filePath = null) {
        this.tmlInput = tmlInput;
        this.filePath = filePath;
        this.logger = require('log4js').getLogger();

        if (filePath) {
            this.loadFromFile();
        }
    }

    loadFromFile() {
        try {
            this.tmlInput = fs.readFileSync(this.filePath, 'utf-8');
        } catch (error) {
            this.logger.error(`Error loading TML file: ${error}`);
            throw error;
        }
    }

    parse() {
        try {
            const namePattern = /<name>(.*?)<\/name>/;
            const descriptionPattern = /<description>(.*?)<\/description>/;
            const paramsPattern = /<params>(.*?)<\/params>/s;

            const nameMatch = this.tmlInput.match(namePattern);
            const descriptionMatch = this.tmlInput.match(descriptionPattern);
            const paramsMatch = this.tmlInput.match(paramsPattern);

            if (!(nameMatch && descriptionMatch && paramsMatch)) {
                throw new Error("Invalid TML input format");
            }

            const name = nameMatch[1].trim();
            const description = descriptionMatch[1].trim();
            const paramsText = paramsMatch[1].trim().split('\n');
            const params = {};
            for (const param of paramsText) {
                const paramParts = param.split(':');
                const paramName = paramParts[0].trim();
                const paramDescription = paramParts[1].trim();
                params[paramName] = paramDescription;
            }

            const output = {
                type: "function",
                function: {
                    name: name,
                    description: description,
                    parameters: {
                        type: "object",
                        properties: Object.fromEntries(Object.entries(params).map(([param, desc]) => [param, { type: "string", description: desc }])),
                        required: Object.keys(params)
                    }
                }
            };

            return output;
        } catch (error) {
            this.logger.error(`An error occurred while parsing TML input: ${error}`);
            throw error;
        }
    }
}

module.exports = TMLParser;

File: parse.py
import re
import json
import logging

class TMLParser:
    def __init__(self, tml_input=None, file_path=None):
        self.tml_input = tml_input
        self.file_path = file_path
        self.logger = logging.getLogger(__name__)

        if file_path:
            self.load_from_file()

    def load_from_file(self):
        try:
            with open(self.file_path, 'r') as file:
                self.tml_input = file.read()
        except Exception as e:
            self.logger.error("Error loading TML file: %s", str(e))
            raise

    def parse(self):
        try:
            name_pattern = re.compile(r'<name>(.*?)<\/name>')
            description_pattern = re.compile(r'<description>(.*?)<\/description>')
            params_pattern = re.compile(r'<params>(.*?)<\/params>', re.DOTALL)

            name_match = name_pattern.search(self.tml_input)
            description_match = description_pattern.search(self.tml_input)
            params_match = params_pattern.search(self.tml_input)

            if not (name_match and description_match and params_match):
                raise ValueError("Invalid TML input format")

            name = name_match.group(1).strip()
            description = description_match.group(1).strip()
            params_text = params_match.group(1).strip()
            params = {}
            params_text = params_text.strip().split('\n')
            for param in params_text:
                param_parts = param.split(':')
                param_name = param_parts[0].strip()
                param_description = param_parts[1].strip()
                params[param_name] = param_description

            output = {
                "type": "function",
                "function": {
                    "name": name,
                    "description": description,
                    "parameters": {
                        "type": "object",
                        "properties": {param: {"type": "string", "description": desc} for param, desc in params.items()},
                        "required": list(params.keys())
                    }
                }
            }

            return output
        except Exception as e:
            self.logger.error("An error occurred while parsing TML input: %s", str(e))
            raise